lib/cuda/cudaBasebandBeamformer.cpp

```diff
@@ -0,0 +1,141 @@
+#include "cudaBasebandBeamformer.hpp"
+
+#include "math.h"
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+
+#define CHECK_CU_ERROR(result)                                                                     \
+    if (result != CUDA_SUCCESS) {                                                                  \
+        const char* errstr = NULL;                                                                 \
+        cuGetErrorString(result, &errstr);                                                         \
+        internal_logging(LOG_ERR, __log_prefix, "Error at {:s}:{:d}; Error type: {:s}", __FILE__,  \
+                         __LINE__, errstr);                                                        \
+        std::abort();                                                                              \
+    }
+
+
+REGISTER_CUDA_COMMAND(cudaBasebandBeamformer);
+
+cudaBasebandBeamformer::cudaBasebandBeamformer(Config& config, const std::string& unique_name,
+                                               bufferContainer& host_buffers,
+                                               cudaDeviceInterface& device) :
+    cudaCommand(config, unique_name, host_buffers, device, "baseband_beamformer", "bb.ptx") {
+    _num_elements = config.get<int>(unique_name, "num_elements");
+    _num_local_freq = config.get<int>(unique_name, "num_local_freq");
+    _samples_per_data_set = config.get<int>(unique_name, "samples_per_data_set");
+    _num_beams = config.get<int>(unique_name, "num_beams");
+    _gpu_mem_voltage = config.get<std::string>(unique_name, "gpu_mem_voltage");
+    _gpu_mem_phase = config.get<std::string>(unique_name, "gpu_mem_phase");
+    _gpu_mem_output_scaling = config.get<std::string>(unique_name, "gpu_mem_output_scaling");
+    _gpu_mem_formed_beams = config.get<std::string>(unique_name, "gpu_mem_formed_beams");
+    _gpu_mem_info = config.get<std::string>(unique_name, "gpu_mem_info");
+
+    command_type = gpuCommandType::KERNEL;
+
+    std::vector<std::string> opts = {
+        "--gpu-name=sm_86",
+        "--verbose",
+    };
+    build_ptx({kernel_name}, opts);
+}
+
+cudaBasebandBeamformer::~cudaBasebandBeamformer() {}
+
+
+// This struct is Erik's interpretation of what Julia is expecting for its "CuDevArray" type.
+template<typename T, int64_t N>
+struct CuDeviceArray {
+    T* ptr;
+    int64_t maxsize;
+    int64_t dims[N];
+    int64_t len;
+};
+typedef CuDeviceArray<int32_t, 1> kernel_arg;
+
+cudaEvent_t cudaBasebandBeamformer::execute(int gpu_frame_id, cudaEvent_t pre_event) {
+    pre_execute(gpu_frame_id);
+
+    size_t voltage_len = (size_t)_num_elements * _num_local_freq * _samples_per_data_set;
```

Comment:
It feels like a really minor optimization, but we could make these `*_len` variables member values, and compute them once in the constructor.  Which might also make the code a little easier to read.

Reply:
Done!